modules/consensus/synchronize.go

```diff
@@ -389,13 +403,18 @@ func (cs *ConsensusSet) rpcRelayHeader(conn modules.PeerConn) error {
 	} else if err != nil {
 		return err
 	}
-	// If the header is valid and extends the heaviest chain, fetch, accept it,
-	// and broadcast it.
+
+	// If the header is valid and extends the heaviest chain, fetch the
+	// corresponding block. Call needs to be made in a separate goroutine
+	// because an exported call to the gateway is used, which is a deadlock
+	// risk given that rpcRelayHeader is called from the gateway.
+	wg.Add(1)
 	go func() {
-		err := cs.gateway.RPC(conn.RPCAddr(), "SendBlk", cs.threadedReceiveBlock(h.ID()))
+		err = cs.gateway.RPC(conn.RPCAddr(), "SendBlk", cs.threadedReceiveBlock(h.ID()))
 		if err != nil {
 			cs.log.Debugln("WARN: failed to get header's corresponding block:", err)
 		}
+		wg.Done()
```

Comment:
Oh, I see what you are saying about the WaitGroup being unnecessary. If we ignore the goroutine that gets spawned in the event of an error, we have a case where a single goroutine is being spawned at the end of execution of the function, meaning you can just add the `tg.Done` inside of the goroutine instead of in a defer statement at the top.

But that's only true because only a single goroutine is spawned, and it gets spawned at the very end of the function. If it was spawned before the end, you'd still need a WaitGroup because you need to know that both the parent and the child thread have completed.

As for the goroutine that gets spawned in the error - I missed it, good catch.

I think I would still prefer to use the `defer go Wait Done` pattern because it applies more broadly. While somewhat unnecessary, it's a recognizable pattern for managing synchronization between functions that spawn multiple goroutines.